### ISSUE TITLE
using minimap2 with samtools-v0.1.19

### ISSUE THREAD

**Issue by sjfleck** [samtools/samtools#1594]
Hello,
I'm running an old version of samtools (v0.1.19) because I'm using an older program that requires v0.1.18 or v0.1.19 called PSMC. My goal is to run this command successfully:
samtools mpileup -C 50 -uf ref.fa aln.bam | bcftools view -c - | vcfutils.pl vcf2fq -d 10 -D 100 | gzip > diploid.fq.gz

I submit .sh scripts to my university's cluster for them to run. 
uname -sr: Linux 3.10.0-1160.49.1.el7.x86_64
uname -m: x86_64
gcc --version: gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)

For the sample I'm working with, I only have Oxford Nanopore long-reads and am attempting to map them using minimap2. I've been reading through the issues tab for minimap2, samtools, and psmc, but I haven't been able to figure this out. Is samtools compatible with minimap2 and/or long-read mapping to a reference? 

I've tried many different options for minimap2 based on some comments I've seen, but here are my most recent commands and options:
minimap2 -t 15 -L --eqx --secondary=no -o aln.mm2.sam -R '@RG\tID:ID.$SAMPLE\tSM:SAMPLE.$SAMPLE' -ax map-ont $GENOME.fasta $ONT 
samtools view -q 25 -h -b -S aln.mm2.sam | samtools sort -@ 16 -m 4G - aln.mm2.sorted
samtools index aln.mm2.sorted.bam
samtools mpileup -d 8000 -C 50 -uf $GENOME.fasta aln.mm2.sorted.bam | bcftools view -cg - | vcfutils.pl vcf2fq | gzip > aln.mm2.sorted.fq.gz

I have 3 samples that are giving 3 different issues:
SAMPLE A:
*** Error in `samtools': corrupted size vs. prev_size: 0x0000000000b2a990 ***

SAMPLE B:
seemingly running normally

SAMPLE C:
Use of uninitialized value $l in numeric lt (<) at /util/academic/samtools/samtools-0.1.19/bcftools/vcfutils.pl line 545, <> line 3616.
Use of uninitialized value $l in numeric lt (<) at /util/academic/samtools/samtools-0.1.19/bcftools/vcfutils.pl line 545, <> line 3616.

There isn't anything different about these samples besides the raw reads (and corresponding reference assemblies) being from different species. All of the commands are identical. 

I know this version of samtools is outdated and vcfutils.pl is no longer supported , but it's my understanding that these tools are necessary for running PSMC. Any help would be greatly appreciated. Thank you.

**Comment by jkbonfield** [comment]
Samtools 0.1.19 is *ancient* and predates long read technology.  It's therefore very unlikely to do a good job of calling variants in it either.  We also don't support this and cannot fix bugs in it.  (Then again, it does appear this samtools release is comtemporary with that RedHat and gcc!)

I would strongly recommend working out how to get PMSC to be compatible with a modern samtools, and perhaps raising an issue on their github repository.  We cannot offer any guidance on this matter ourselves.

**Comment by sjfleck** [comment]
Thank you for your quick response. I have another quick question. Previously, using samtools-v0.1.19, I was able to run this command:
samtools mpileup -C 50 -uf ref.fa aln.bam | bcftools view -c - | vcfutils.pl vcf2fq -d 10 -D 100 | gzip > diploid.fq.gz

because vcfutils.pl is no longer supported, I'm unaware of a way to convert the VCF into a fastq file. Does the newer version have this functionality in some way? Currently, I'm only capable of generating a VCF file with this bcftools command (since mpileup moved from samtools to bcftools):
bcftools mpileup -C 50 -Ou --threads 16 -f ref.fa aln.sorted.bam | bcftools call -c -Ov -o aln.sorted.pdrm.original_call.vcf

I'm also now realizing that I should probably take this to the bcftools issue section, but I thought I would still ask since they're both under samtools. Thank you again

**Comment by lcanesin** [comment]
Hi @sjfleck!

Did you get this to work? I'm trying to use PSMC with HiFi-CCS reads and minimap2. But obviously with no success so far. 
Gonna try now your tips with bcftools.

**Comment by sjfleck** [comment]
I was able to get it to work with ONT long-reads and it seemed pretty consistent with running PSMC with short reads as well. You can see what was said by Heng Li here:
https://github.com/lh3/psmc/issues/41

And you can see how I ended up running PSMCR here:
https://github.com/emmanuelparadis/psmcr/issues/7

**Comment by lcanesin** [comment]
That's great! 
I followed your tips and got farther than I did on my own. But I am still confused with mapping parameters and the usage of the most recent tools, namely using bcftools to perform the snp calling. What parameters did you use in minimap2 run? I just need to set the -L flag? Did you use the samtools 0.19 or the most recent version?

Thank you! 
I hope @lh3 can update PSMC to handle long read data. It would be great to combine PSMC and the ongoing genome projects around the world.
